File: java_sim/simulation_188623/src/Zwierze.java
public class Zwierze extends Organizm {

    Zwierze(Swiat swiatOrganizmu, int sila, int inicjatywa, int x_pos, int y_pos, int wiek, boolean czyCzeka) {
        super(swiatOrganizmu, sila, inicjatywa, x_pos, y_pos, wiek, czyCzeka);
    }

    boolean czyTenSamGatunek(Organizm kolidujacyOrganizm){
        return (this.getNazwa() == kolidujacyOrganizm.getNazwa());
    }

    void rozmnazanie(){
        int newPosition[] = znajdzKolejneWolnePole();

        if(newPosition[0] == -1 && newPosition[1] == -1){
            System.out.println("Rozmnazanie "+this.getNazwa()+" nieudane, brak dostepnych pol");
            return;
        }

        klonowanie(newPosition[0],newPosition[1]);
    }

    void walka(Organizm oponent){
        this.komunikat("[Walka]", oponent);

        if (sila >= oponent.getSila() && !(oponent instanceof BarszczSosnowskiego)) {

            System.out.println("Zwyciestwo: "+ this.getNazwa()); 
            Komentator.DodajKomentarz("Zwyciestwo: "+ this.getNazwa());   
    
            if (oponent instanceof Guarana) {

                System.out.println(this.getNazwa()+" ("+koordynaty[0]+","+koordynaty[1]+") sila zwiekszona");
                Komentator.DodajKomentarz(this.getNazwa()+" ("+koordynaty[0]+","+koordynaty[1]+") sila zwiekszona");             
                this.sila += 3;
            }
    
            swiatOrganizmu.plansza[koordynaty[0]][koordynaty[1]] = null;
            koordynaty[0] = oponent.getX();
            koordynaty[1] = oponent.getY();
            if(this instanceof Czlowiek) {
                this.swiatOrganizmu.setPlayerX(koordynaty[0]);
                this.swiatOrganizmu.setPlayerY(koordynaty[1]);
            }
            swiatOrganizmu.plansza[koordynaty[0]][koordynaty[1]] = this;
            swiatOrganizmu.usunOrganizm(oponent);
        }
        else {
            System.out.println("Przegrana: "+this.getNazwa());         
            swiatOrganizmu.plansza[koordynaty[0]][koordynaty[1]] = null;
            swiatOrganizmu.usunOrganizm(this);
        }
    }

    void kolizja(Organizm kolidujacyOrganizm){
        if(czyTenSamGatunek(kolidujacyOrganizm)){
            if(kolidujacyOrganizm instanceof Czlowiek){
                return;
            }

            this.komunikat("[Rozmnazanie]", kolidujacyOrganizm);
            rozmnazanie();
            return;
        }
        else{
            if(kolidujacyOrganizm.odbicie(this)){
                this.komunikat("[Odbicie]", kolidujacyOrganizm);
                return;
            }
    
            this.walka(kolidujacyOrganizm);  
        }
    }

    @Override
    Organizm klonowanie(int posX, int posY){
        return null;
    }

    @Override
    String getNazwa()
    {
        return null;
    }

    @Override
    char getZnaczek() {
        return '-'; 
    }

    @Override
    void akcja() {
        //szansa zolwia na ominiecie ruchu
        if(this.getNazwa().equals("Zolw")){
            wiek++;
        }
        int newPosition[] = znajdzNowePole();

        if(swiatOrganizmu.plansza[newPosition[0]][newPosition[1]] == null){
            swiatOrganizmu.plansza[newPosition[0]][newPosition[1]] = this;
            swiatOrganizmu.plansza[koordynaty[0]][koordynaty[1]] = null;
            koordynaty[0] = newPosition[0];
            koordynaty[1] = newPosition[1];    
        }
        else{
            kolizja(swiatOrganizmu.plansza[newPosition[0]][newPosition[1]]);
        }
    }
    
}

File: cpp_sim/188623/Zwierze.cpp
#include "Zwierze.h"
#include "BarszczSosnowskiego.h"
#include "Guarana.h"
#include "Czlowiek.h"
#include <typeinfo>

Zwierze::Zwierze(Swiat *swiatOrganizmu, int sila, int initiative, int posX, int posY, int age, bool czyCzeka)
: Organizm(swiatOrganizmu, sila, initiative, posX, posY, age, czyCzeka){
};

bool Zwierze::czyTenSamGatunek(Organizm &entity){
    return (this->getNazwa() == entity.getNazwa());
};

void Zwierze::rozmnazanie(){

    int *newPosition = znajdzKolejneWolnePole();

    if(newPosition[0] == -1 && newPosition[1] == -1){
        std::cout << "Rozmnazanie " << this->getNazwa() << " nieudane, brak dostepnych pol\n";
        return;
    }

    klonowanie(newPosition[0], newPosition[1]);
    delete newPosition;
};

void Zwierze::akcja(){

    //szansa zolwia na ominiecie ruchu
    if(this->getNazwa().compare("Zolw")){
        wiek++;
    }
    int *newPosition = znajdzNowePole();
    
    if(swiatOrganizmu->plansza[newPosition[0]][newPosition[1]] == NULL){
        swiatOrganizmu->plansza[newPosition[0]][newPosition[1]] = this;
        swiatOrganizmu->plansza[koordynaty[0]][koordynaty[1]] = NULL;
        koordynaty[0] = newPosition[0];
        koordynaty[1] = newPosition[1];
    }
    else{
    
        kolizja(swiatOrganizmu->plansza[newPosition[0]][newPosition[1]]);
    }
    delete newPosition;
};

void Zwierze::walka(Organizm* oponent) {
    this->komunikat("[Walka]", oponent);

    if (sila >= oponent->getSila() && typeid(*oponent).hash_code() != typeid(BarszczSosnowskiego).hash_code()) {

        std::cout << "Zwyciestwo: " << this->getNazwa() << "\n";

        if (typeid(*oponent).hash_code() == typeid(Guarana).hash_code()) {
            std::cout << this->getNazwa() << " (" << koordynaty[0] << "," << koordynaty[1] << ") sila zwiekszona\n";
            sila += 3;
        }

        swiatOrganizmu->plansza[koordynaty[0]][koordynaty[1]] = NULL;
        koordynaty[0] = oponent->getX();
        koordynaty[1] = oponent->getY();
        swiatOrganizmu->plansza[koordynaty[0]][koordynaty[1]] = this;

        swiatOrganizmu->podgladOrganizmow->usunOrganizm(oponent);
    }
    else {
        std::cout << "Przegrana: " << this->getNazwa() << "\n";
        swiatOrganizmu->plansza[koordynaty[0]][koordynaty[1]] = NULL;
        swiatOrganizmu->podgladOrganizmow->usunOrganizm(this);
    }
};

void Zwierze::kolizja(Organizm *kolidujacyOrganizm){
    if(czyTenSamGatunek(*kolidujacyOrganizm)){
        
        if (typeid(*kolidujacyOrganizm).hash_code() == typeid(Czlowiek).hash_code()) {
            return;
        }

        this->komunikat("[Rozmnazanie]", kolidujacyOrganizm);
        
        rozmnazanie();
        return;
    }
    else{
        
        if(kolidujacyOrganizm->odbicie(this)){
            this->komunikat("[Odbicie]", kolidujacyOrganizm);
            return;
        }

        this->walka(kolidujacyOrganizm);        
    }
};

Zwierze::~Zwierze(){
};